Release v18.17.0a1

roycaihw opened the pull request:
- Updated submodule
- Cherrypicked important changes in master branch since f0dbe3897017ae2cb7e89dde3358816e9aa2a37b: commit 3-10. See my reason below.
- Updated version constant v18.17.0a1
- Added a temporary workaround for the required env LIBRARY https://github.com/kubernetes-client/gen/pull/183
- Generated the client, removed the test directory, and cherry-picked changes from https://github.com/kubernetes-client/python/pull/1317 following the [hot issues](https://github.com/kubernetes-client/python/blob/master/devel/release.md#hot-issues) instruction. 
- Updated CHANGELOG and README.

I found it quite hard to resolve the conflicts when pulling master into the release branch. The conflict was caused by we generating v17 and v18 snapshots in parallel. As a workaround I cherrypicked important changes in the master branch into the release branch.

yliaog commented on scripts/update-client.sh:
```diff
@@ -42,6 +42,9 @@ SETTING_FILE="${TEMP_FOLDER}/settings"
 echo "export PACKAGE_NAME=\"client\"" >> $SETTING_FILE
+# Workaround for https://github.com/kubernetes-client/gen/pull/183.
+# The env is required by not used by python client.
+echo "export LIBRARY=\"undefined\"" >> $SETTING_FILE
```
should this be added to 'master' branch?

roycaihw replied:
hopefully not once https://github.com/kubernetes-client/gen/pull/187 is merged. I filed https://github.com/kubernetes-client/python/issues/1412 to track documenting the issue.

yliaog replied:
s/The env is required by not used by python client/The env is required but not used by python client/

roycaihw replied:
done

yliaog commented:
v17 and v18 snapshots are in different branches, why would that cause conflicts when pulling from master into release branch?

roycaihw commented:
> why would that cause conflicts when pulling from master into release branch

I think it's because the 18.0 snapshot is not in the master branch, and the master branch has the 17.0 snapshot. 

Pulling master from `upstream/release-18.0`:
```
$ git log
commit 7a0e799a028dc5581c18fa399d45ff7af7c9d02e (HEAD -> release-18.0-bak, upstream/release-18.0)
$ git fetch upstream
$ git pull upstream master
```
An example conflict:
```
$ git diff scripts/constants.py
diff --cc scripts/constants.py
index 6e30bd0c6,084911e92..000000000
--- a/scripts/constants.py
+++ b/scripts/constants.py
@@@ -15,10 -15,10 +15,17 @@@
  import sys

  # Kubernetes branch to get the OpenAPI spec from.
++<<<<<<< HEAD
 +KUBERNETES_BRANCH = "release-1.18"
 +
 +# client version for packaging and releasing.
 +CLIENT_VERSION = "18.0.0-snapshot"
++=======
+ KUBERNETES_BRANCH = "release-1.17"
+
+ # client version for packaging and releasing.
+ CLIENT_VERSION = "17.0.0-snapshot"
++>>>>>>> b839eee939455d17b99f1335f6318732c6bea77c
...
```
git thinks a lot of files are modified by both branches:
```
$ git status
On branch release-18.0-bak
You have unmerged paths.
  (fix conflicts and run "git commit")
  (use "git merge --abort" to abort the merge)

Changes to be committed:
	modified:   .github/workflows/test.yaml
	modified:   .travis.yml
	modified:   examples/pod_exec.py
	modified:   examples/remote_cluster.py
	modified:   kubernetes/base
	new file:   kubernetes/e2e_test/port_server.py
	modified:   kubernetes/e2e_test/test_client.py
	modified:   kubernetes/e2e_test/test_watch.py
	new file:   kubernetes/leaderelection
	modified:   kubernetes/utils/create_from_yaml.py
	modified:   tox.ini

Unmerged paths:
  (use "git add/rm <file>..." as appropriate to mark resolution)
	both modified:   CHANGELOG.md
	both modified:   kubernetes/.openapi-generator/swagger.json.sha256
	both modified:   kubernetes/README.md
	both modified:   kubernetes/__init__.py
	both modified:   kubernetes/client/__init__.py
	both modified:   kubernetes/client/api/admissionregistration_api.py
	both modified:   kubernetes/client/api/admissionregistration_v1_api.py
	both modified:   kubernetes/client/api/admissionregistration_v1beta1_api.py
	both modified:   kubernetes/client/api/apiextensions_api.py
	both modified:   kubernetes/client/api/apiextensions_v1_api.py
	both modified:   kubernetes/client/api/apiextensions_v1beta1_api.py
	both modified:   kubernetes/client/api/apiregistration_api.py
	both modified:   kubernetes/client/api/apiregistration_v1_api.py
	both modified:   kubernetes/client/api/apiregistration_v1beta1_api.py
	both modified:   kubernetes/client/api/apis_api.py
	both modified:   kubernetes/client/api/apps_api.py
	both modified:   kubernetes/client/api/apps_v1_api.py
	deleted by us:   kubernetes/client/api/apps_v1beta1_api.py
	deleted by us:   kubernetes/client/api/apps_v1beta2_api.py
	both modified:   kubernetes/client/api/auditregistration_api.py
	both modified:   kubernetes/client/api/auditregistration_v1alpha1_api.py
	both modified:   kubernetes/client/api/authentication_api.py
	both modified:   kubernetes/client/api/authentication_v1_api.py
	both modified:   kubernetes/client/api/authentication_v1beta1_api.py
	both modified:   kubernetes/client/api/authorization_api.py
	both modified:   kubernetes/client/api/authorization_v1_api.py
	both modified:   kubernetes/client/api/authorization_v1beta1_api.py
	both modified:   kubernetes/client/api/autoscaling_api.py
	both modified:   kubernetes/client/api/autoscaling_v1_api.py
	both modified:   kubernetes/client/api/autoscaling_v2beta1_api.py
	both modified:   kubernetes/client/api/autoscaling_v2beta2_api.py
	both modified:   kubernetes/client/api/batch_api.py
	both modified:   kubernetes/client/api/batch_v1_api.py
	both modified:   kubernetes/client/api/batch_v1beta1_api.py
	both modified:   kubernetes/client/api/batch_v2alpha1_api.py
	both modified:   kubernetes/client/api/certificates_api.py
	both modified:   kubernetes/client/api/certificates_v1beta1_api.py
	both modified:   kubernetes/client/api/coordination_api.py
	both modified:   kubernetes/client/api/coordination_v1_api.py
	both modified:   kubernetes/client/api/coordination_v1beta1_api.py
	both modified:   kubernetes/client/api/core_api.py
	both modified:   kubernetes/client/api/core_v1_api.py
	both modified:   kubernetes/client/api/custom_objects_api.py
	both modified:   kubernetes/client/api/discovery_api.py
	both modified:   kubernetes/client/api/discovery_v1beta1_api.py
	both modified:   kubernetes/client/api/events_api.py
	both modified:   kubernetes/client/api/events_v1beta1_api.py
	both modified:   kubernetes/client/api/extensions_api.py
	both modified:   kubernetes/client/api/extensions_v1beta1_api.py
	both added:      kubernetes/client/api/flowcontrol_apiserver_api.py
	both added:      kubernetes/client/api/flowcontrol_apiserver_v1alpha1_api.py
	both modified:   kubernetes/client/api/logs_api.py
	both modified:   kubernetes/client/api/networking_api.py
	both modified:   kubernetes/client/api/networking_v1_api.py
	both modified:   kubernetes/client/api/networking_v1beta1_api.py
	both modified:   kubernetes/client/api/node_api.py
	both modified:   kubernetes/client/api/node_v1alpha1_api.py
	both modified:   kubernetes/client/api/node_v1beta1_api.py
	both modified:   kubernetes/client/api/policy_api.py
	both modified:   kubernetes/client/api/policy_v1beta1_api.py
	both modified:   kubernetes/client/api/rbac_authorization_api.py
	both modified:   kubernetes/client/api/rbac_authorization_v1_api.py
	both modified:   kubernetes/client/api/rbac_authorization_v1alpha1_api.py
	both modified:   kubernetes/client/api/rbac_authorization_v1beta1_api.py
	both modified:   kubernetes/client/api/scheduling_api.py
	both modified:   kubernetes/client/api/scheduling_v1_api.py
	both modified:   kubernetes/client/api/scheduling_v1alpha1_api.py
	both modified:   kubernetes/client/api/scheduling_v1beta1_api.py
	both modified:   kubernetes/client/api/settings_api.py
	both modified:   kubernetes/client/api/settings_v1alpha1_api.py
	both modified:   kubernetes/client/api/storage_api.py
	both modified:   kubernetes/client/api/storage_v1_api.py
	both modified:   kubernetes/client/api/storage_v1alpha1_api.py
	both modified:   kubernetes/client/api/storage_v1beta1_api.py
...
```

yliaog commented:
the current release process does not support multiple parallel releasing well, better to document your release process for v18, so v19 and v20 can follow the same steps.

yliaog commented:
/lgtm
/hold

roycaihw commented:
Filed https://github.com/kubernetes-client/python/issues/1414 to track documenting the release process. I will try getting to it today or tomorrow.

yliaog commented:
/lgtm

roycaihw commented:
/hold cancel

the documentation tasks don't block this PR
